packages/integration/src/ScmIntegrations.ts

```diff
@@ -21,27 +21,64 @@ import { GitHubIntegration } from './github/GitHubIntegration';
 import { GitLabIntegration } from './gitlab/GitLabIntegration';
 import {
   ScmIntegration,
-  ScmIntegrationPredicateTuple,
   ScmIntegrationRegistry,
+  ScmIntegrationsGroup,
 } from './types';
 
+type IntegrationsByType = {
+  azure: ScmIntegrationsGroup<AzureIntegration>;
+  bitbucket: ScmIntegrationsGroup<BitbucketIntegration>;
+  github: ScmIntegrationsGroup<GitHubIntegration>;
+  gitlab: ScmIntegrationsGroup<GitLabIntegration>;
+};
+
 export class ScmIntegrations implements ScmIntegrationRegistry {
+  private readonly byType: IntegrationsByType;
+
   static fromConfig(config: Config): ScmIntegrations {
-    return new ScmIntegrations([
-      ...AzureIntegration.factory({ config }),
-      ...BitbucketIntegration.factory({ config }),
-      ...GitHubIntegration.factory({ config }),
-      ...GitLabIntegration.factory({ config }),
-    ]);
+    return new ScmIntegrations({
+      azure: AzureIntegration.factory({ config }),
+      bitbucket: BitbucketIntegration.factory({ config }),
+      github: GitHubIntegration.factory({ config }),
+      gitlab: GitLabIntegration.factory({ config }),
+    });
+  }
+
+  constructor(integrationsByType: IntegrationsByType) {
+    this.byType = integrationsByType;
   }
 
-  constructor(private readonly integrations: ScmIntegrationPredicateTuple[]) {}
+  get azure(): ScmIntegrationsGroup<AzureIntegration> {
+    return this.byType.azure;
+  }
+
+  get bitbucket(): ScmIntegrationsGroup<BitbucketIntegration> {
+    return this.byType.bitbucket;
+  }
+
+  get github(): ScmIntegrationsGroup<GitHubIntegration> {
+    return this.byType.github;
+  }
+
+  get gitlab(): ScmIntegrationsGroup<GitLabIntegration> {
+    return this.byType.gitlab;
+  }
 
   list(): ScmIntegration[] {
-    return this.integrations.map(i => i.integration);
+    return Object.values(this.byType).flatMap(
+      i => i.list() as ScmIntegration[],
+    );
+  }
+
+  byUrl(url: string | URL): ScmIntegration | undefined {
+    return Object.values(this.byType)
+      .map(i => i.byUrl(url))
+      .find(Boolean);
```

Comment:
`.find(i => i.byUrl(url))`? 😅

Reply:
Oh, uh, ofc 😛

Reply:
No hang on, that's not the same thing. Your code returns the first `i` for which the function returns truthy, my code returns the first `i.byUrl()` **result** that's truthy.

Reply:
Added a test to exercise all of this as well